File: lib/MESCIUS/ComponentOne/WinForms/C1TouchToolKit/PolicySourceCodes/C1GanttViewZoomPolicy.vb
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Text
Imports C1.Win.C1GanttView
Imports GrapeCity.Win.MultiTouch

Public Class C1GanttViewZoomPolicy
	Inherits ZoomPolicy
	Public Overrides ReadOnly Property TargetType() As Type
		Get
			Return GetType(C1.Win.C1GanttView.C1GanttView)
		End Get
	End Property

	Private _ganttViewGridWidthCache As New Dictionary(Of C1GanttView, Integer)()

	Public Overrides Sub ZoomBounds(control As System.Windows.Forms.Control, infos As ZoomBoundsInfo)
		Dim ganttView As C1GanttView = TryCast(control, C1GanttView)
		SetGanttviewWidth(ganttView, infos.Zoom(GetGanttViewGridWidth(ganttView)))
		ganttView.Timescale.BottomTier.MinWidth = infos.Zoom(ganttView.Timescale.BottomTier.MinWidth)
		ganttView.Timescale.TopTier.MinWidth = infos.Zoom(ganttView.Timescale.TopTier.MinWidth)
		ganttView.Timescale.MiddleTier.MinWidth = infos.Zoom(ganttView.Timescale.MiddleTier.MinWidth)
		For Each column As BaseColumn In ganttView.Columns
			If column.Width > 0 Then
				column.Width = infos.Zoom(column.Width)
			End If
		Next
		MyBase.ZoomBounds(control, infos)
	End Sub

	Private Sub SetGanttviewWidth(ganttView As C1GanttView, value As Integer)
		_ganttViewGridWidthCache(ganttView) = value
		ganttView.GridWidth = value
	End Sub

	Private Function GetGanttViewGridWidth(ganttView As C1GanttView) As Integer
		Dim value As Integer
		If _ganttViewGridWidthCache.TryGetValue(ganttView, value) AndAlso (value = ganttView.GridWidth OrElse (value < 80 AndAlso ganttView.GridWidth = 80)) Then
			Return value
		Else
			_ganttViewGridWidthCache(ganttView) = ganttView.GridWidth
			Return ganttView.GridWidth
		End If
	End Function
End Class

File: lib/MESCIUS/ComponentOne/WinForms/C1TouchToolKit/PolicySourceCodes/C1GanttViewZoomPolicy.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Text;
using C1.Win.C1GanttView;
using GrapeCity.Win.MultiTouch;

namespace GrapeCity.Win.MultiTouch.ZoomPolicys
{
    public class C1GanttViewZoomPolicy : ZoomPolicy
    {
        public override Type TargetType
        {
            get { return typeof(C1.Win.C1GanttView.C1GanttView); }
        }

        private Dictionary<C1GanttView, int> _ganttViewGridWidthCache = new Dictionary<C1GanttView, int>();

        public override void ZoomBounds(System.Windows.Forms.Control control, ZoomBoundsInfo infos)
        {
            C1GanttView ganttView = control as C1GanttView;
            SetGanttviewWidth(ganttView,infos.Zoom(GetGanttViewGridWidth(ganttView)));
            ganttView.Timescale.BottomTier.MinWidth = infos.Zoom(ganttView.Timescale.BottomTier.MinWidth);
            ganttView.Timescale.TopTier.MinWidth = infos.Zoom(ganttView.Timescale.TopTier.MinWidth);
            ganttView.Timescale.MiddleTier.MinWidth = infos.Zoom(ganttView.Timescale.MiddleTier.MinWidth);
            foreach (BaseColumn column in ganttView.Columns)
            {
                if (column.Width > 0)
                {
                    column.Width = infos.Zoom(column.Width);
                }
            }
            base.ZoomBounds(control, infos);
        }

        private void SetGanttviewWidth(C1GanttView ganttView, int value)
        {
            _ganttViewGridWidthCache[ganttView] = value;
            ganttView.GridWidth = value;
        }

        private int GetGanttViewGridWidth(C1GanttView ganttView)
        {
            int value;
            if (_ganttViewGridWidthCache.TryGetValue(ganttView, out value) && (value == ganttView.GridWidth || (value < 80 && ganttView.GridWidth == 80)))
            {
                return value;
            }
            else
            {
                _ganttViewGridWidthCache[ganttView] = ganttView.GridWidth;
                return ganttView.GridWidth;
            }
        }
    }
}
